File: src/chain-observers/factories/factory-base.ts
import type { WorkerBee } from "../../bot";
import { WorkerBeeError } from "../../errors";
import { createFactoryCircularDependencyErrorMessage, createFactoryUnsupportedClassifierErrorMessage } from "../../utils/error-helper";
import { DynamicGlobalPropertiesClassifier } from "../classifiers";
import { CollectorClassifierBase, IEvaluationContextClass, TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { CollectorBase } from "../collectors/collector-base";
import { ObserverMediator } from "../observer-mediator";
import { DataEvaluationContext } from "./data-evaluation-context";

export enum EClassifierOrigin {
  FILTER = "filter",
  PROVIDER = "provider",
  FACTORY = "factory"
}

export type AnyCollectorClass = new (...args: any[]) => CollectorBase<CollectorClassifierBase<any, any, any, any, any>>;

export class FactoryBase {
  protected collectors = new Map<AnyCollectorClass, CollectorBase<CollectorClassifierBase<any, any, any, any, any>>>();
  protected collectorsPerClassifier = new Map<IEvaluationContextClass, AnyCollectorClass>();
  protected currentBlockNumber?: number;

  public constructor(
    protected readonly worker: WorkerBee
  ) {}

  private readonly timings: Record<string, number> = {};
  private lastStart = Date.now();

  public getTimings(): Readonly<Record<string, number>> {
    this.timings.total = Date.now() - this.lastStart;
    return this.timings;
  }

  public addTiming(name: string, time: number) {
    this.timings[name] = (this.timings[name] ?? 0) + time;
  }

  /**
   * Should be called to register a classifier with a collector.
   * If the collector is already registered for the classifier, or not yet registered, it will return false.
   * If the collector is different, it will replace the existing one and return true.
   */
  public registerClassifier<
    Classifier extends IEvaluationContextClass,
    Collector extends AnyCollectorClass
  >(classifier: Classifier, collector: Collector, ...constructorArgs: ConstructorParameters<Collector>): boolean {
    const existingCollector = this.collectorsPerClassifier.get(classifier);
    if (existingCollector === collector)
      return false; // Collector already registered for this classifier and is the same class

    /*
     * If the collector is already registered for a different classifier,
     * we need to assign the reference of the existing collector instance to the new classifier to use the same data between classifiers
     * This is useful for cases when the same collector provides data for multiple classifiers
     */
    const collectorInstance = this.collectors.get(collector);
    if (collectorInstance !== undefined) {
      this.collectorsPerClassifier.set(classifier, collector);
      this.collectors.set(collector, collectorInstance);
      return false;
    }

    this.collectorsPerClassifier.set(classifier, collector);
    this.collectors.set(collector, new collector(...constructorArgs));

    return existingCollector === collector;
  }

  /**
   * Should be called to unregister a classifier from a collector.
   * If the collector is not registered for the classifier, it will return false.
   * If the collector is registered, it will remove it and return true.
   */
  public unregisterClassifier(classifier: IEvaluationContextClass): boolean {
    const collector = this.collectorsPerClassifier.get(classifier);
    if (collector === undefined)
      return false; // No collector registered for this classifier

    this.collectorsPerClassifier.delete(classifier);

    const collectorInstance = this.collectors.get(collector);
    if (collectorInstance === undefined)
      return false; // No collector instance registered for this classifier

    this.collectors.delete(collector);

    return true; // Successfully unregistered the collector
  }

  /**
   * Should be called to extend the factory with another factory's collectors.
   * It will register all collectors from the other factory that are not already registered in this factory.
   */
  public extend(other: FactoryBase): void {
    for(const [classifier, otherCollector] of other.collectorsPerClassifier) {
      const thisCollector = this.collectorsPerClassifier.get(classifier);

      if (thisCollector === undefined)
        continue; // Not registered in this factory, so it is unsupported - we will not register it

      // Already registered, just update the instance reference to use collected data from the other factory
      if (thisCollector === otherCollector) {
        const existingOtherCollectorInstance = other.collectors.get(otherCollector);
        if (existingOtherCollectorInstance === undefined)
          throw new WorkerBeeError(
            `Internal error: Collector instance not found for classifier ${classifier.name} in factory ${(other as any).__proto__.constructor.name}`
          );

        this.collectors.set(thisCollector, existingOtherCollectorInstance);
      }
    }
  }

  /**
   * Called by mediator before any notification processing begins for a data collection cycle.
   * This method determines whether the factory should proceed with processing the current cycle.
   * If this method returns true, filters, providers, callbacks and {@link postNotify} will be executed normally.
   * If this method returns false, the entire processing cycle will be skipped, including filters, providers, callbacks and {@link postNotify}.
   */
  public async preNotify(context: DataEvaluationContext, _mediator: ObserverMediator): Promise<boolean> {
    // If this is the first run (no previous block processed), we should proceed
    const dgp = await context.get(DynamicGlobalPropertiesClassifier);

    /*
     * If the current head block is different from the last processed block, we should proceed
     * Will be also true for the first run as currentBlockNumber is undefined
     */
    const hasBlockNumberChanged = this.currentBlockNumber !== dgp.headBlockNumber;

    if (hasBlockNumberChanged)
      this.currentBlockNumber = this.currentBlockNumber ? this.currentBlockNumber + 1 : dgp.headBlockNumber;

    // If we've already processed this block, skip processing
    return hasBlockNumberChanged;
  }

  /**
   * Called by mediator after all notification processing has completed for a data collection cycle.
   * This method provides an opportunity to perform cleanup, finalization, or post-processing
   * tasks after all observers have been notified.
   * Note: Getting inside this method does not mean all filters and callbacks has been processed (ended).
   */
  public async postNotify(_context: DataEvaluationContext, _mediator: ObserverMediator): Promise<void> {}

  public pushClassifier(classifier: TRegisterEvaluationContext, origin: EClassifierOrigin, stack: IEvaluationContextClass[] = []): void {
    const classifierClass = "class" in classifier ? classifier.class : classifier;

    if (stack.includes(classifierClass))
      throw new WorkerBeeError(createFactoryCircularDependencyErrorMessage((this as any).__proto__.constructor.name, classifierClass, origin, stack));

    stack.push(classifierClass);

    const collector = this.collectorsPerClassifier.get(classifierClass);
    if (collector === undefined)
      throw new WorkerBeeError(createFactoryUnsupportedClassifierErrorMessage((this as any).__proto__.constructor.name, classifierClass, origin, stack));

    const instance = this.collectors.get(collector);
    if (instance === undefined)
      throw new WorkerBeeError(
        `Internal error: Collector instance not found for classifier ${classifierClass.name} in factory ${(this as any).__proto__.constructor.name}`
      );

    instance.register("options" in classifier ? classifier.options : undefined);

    for(const dependency of instance.usedContexts()) // Rewrite stack to avoid detecting false circular dependencies when on the same nested level
      this.pushClassifier(dependency, origin, [...stack]);
  }

  public popClassifier(classifier: TRegisterEvaluationContext, origin: EClassifierOrigin, stack: IEvaluationContextClass[] = []): void {
    const classifierClass = "class" in classifier ? classifier.class : classifier;

    if (stack.includes(classifierClass))
      throw new WorkerBeeError(createFactoryCircularDependencyErrorMessage((this as any).__proto__.constructor.name, classifierClass, origin, stack));

    stack.push(classifierClass);

    const collector = this.collectorsPerClassifier.get(classifierClass);
    if (collector === undefined)
      throw new WorkerBeeError(createFactoryUnsupportedClassifierErrorMessage((this as any).__proto__.constructor.name, classifierClass, origin, stack));

    const instance = this.collectors.get(collector);
    if (instance === undefined)
      throw new WorkerBeeError(
        `Internal error: Collector instance not found for classifier ${classifierClass.name} in factory ${(this as any).__proto__.constructor.name}`
      );

    instance.unregister("options" in classifier ? classifier.options : undefined);

    for(const dependency of instance.usedContexts()) // Rewrite stack to avoid detecting false circular dependencies when on the same nested level
      this.popClassifier(dependency, origin, [...stack]);
  }

  private rebuildDataEvaluationContext(): DataEvaluationContext {
    const context = new DataEvaluationContext(this);

    for(const [classifier, collector] of this.collectorsPerClassifier) {
      const collectorInstance = this.collectors.get(collector);
      if (collectorInstance === undefined)
        throw new WorkerBeeError(
          `Internal error: Collector instance not found for classifier ${classifier.name} in factory ${(this as any).__proto__.constructor.name}`
        );

      if (!collectorInstance.hasRegistered) // Ignore collectors that have no registered classifiers
        continue;

      // Dependencies are already pushed by the classifier

      context.inject(classifier, collectorInstance);
    }

    return context;
  }

  public collect(): DataEvaluationContext {
    return this.rebuildDataEvaluationContext();
  }
}

File: python/workerbee/chain_observers/factories/factory_base.py
"""FactoryBase — manages classifier→collector registration and lifecycle.

Mirrors src/chain-observers/factories/factory-base.ts.
"""

from __future__ import annotations

import time
from typing import TYPE_CHECKING, Any

from ..classifiers.collector_classifier_base import CollectorClassifierBase, TRegisterEvaluationContext
from ..collectors.collector_base import CollectorBase
from .data_evaluation_context import DataEvaluationContext

if TYPE_CHECKING:
    from ..interfaces import IWorkerBee
    from ..observer_mediator import ObserverMediator


class EClassifierOrigin:
    FILTER = "filter"
    PROVIDER = "provider"
    FACTORY = "factory"


class FactoryBase:
    def __init__(self, worker: IWorkerBee) -> None:
        self.worker = worker
        self._collectors: dict[type[CollectorBase], CollectorBase] = {}
        self._collectors_per_classifier: dict[type[CollectorClassifierBase[Any]], type[CollectorBase]] = {}
        self._current_block_number: int | None = None
        self._timings: dict[str, float] = {}
        self._last_start: float = time.monotonic()

    def get_timings(self) -> dict[str, float]:
        self._timings["total"] = (time.monotonic() - self._last_start) * 1000
        return dict(self._timings)

    def record_timing(self, name: str, timing: float) -> None:
        """Accumulate ``timing`` (ms) under ``name``. Low-level store; callers
        time blocks via the :meth:`DataEvaluationContext.add_timing` context
        manager rather than computing elapsed time by hand."""
        self._timings[name] = self._timings.get(name, 0) + timing

    def register_classifier(
        self,
        classifier: type[CollectorClassifierBase[Any]],
        collector_class: type[CollectorBase],
        *constructor_args: Any,
    ) -> bool:
        existing = self._collectors_per_classifier.get(classifier)
        if existing is collector_class:
            return False

        instance = self._collectors.get(collector_class)
        if instance is not None:
            self._collectors_per_classifier[classifier] = collector_class
            return False

        self._collectors_per_classifier[classifier] = collector_class
        self._collectors[collector_class] = collector_class(*constructor_args)
        return existing is collector_class

    def unregister_classifier(self, classifier: type[CollectorClassifierBase[Any]]) -> bool:
        collector_class = self._collectors_per_classifier.get(classifier)
        if collector_class is None:
            return False

        del self._collectors_per_classifier[classifier]

        instance = self._collectors.get(collector_class)
        if instance is None:
            return False

        del self._collectors[collector_class]
        return True

    def extend(self, other: FactoryBase) -> None:
        for classifier, other_collector_class in other._collectors_per_classifier.items():
            this_collector_class = self._collectors_per_classifier.get(classifier)
            if this_collector_class is None:
                continue
            if this_collector_class is other_collector_class:
                other_instance = other._collectors.get(other_collector_class)
                if other_instance is None:
                    raise RuntimeError(
                        f"Internal error: Collector instance not found for classifier {classifier.__name__} in factory {type(other).__name__}",
                    )
                self._collectors[this_collector_class] = other_instance

    def push_classifier(
        self,
        classifier: TRegisterEvaluationContext,
        origin: str,
        stack: list[type[CollectorClassifierBase[Any]]] | None = None,
    ) -> None:
        if stack is None:
            stack = []

        classifier_class: type[CollectorClassifierBase[Any]]
        options: dict[str, Any] | None = None

        if isinstance(classifier, dict):
            classifier_class = classifier["class"]
            options = classifier.get("options")
        else:
            classifier_class = classifier

        if classifier_class in stack:
            raise RuntimeError(
                f"Circular dependency detected in factory {type(self).__name__} "
                f"for classifier {classifier_class.__name__} (origin: {origin}, "
                f"stack: {[c.__name__ for c in stack]})",
            )

        stack.append(classifier_class)

        collector_class = self._collectors_per_classifier.get(classifier_class)
        if collector_class is None:
            raise RuntimeError(
                f'Factory "{type(self).__name__}" does not support classifier '
                f'"{classifier_class.__name__}" (origin: {origin}, '
                f"stack: {[c.__name__ for c in stack]})",
            )

        instance = self._collectors.get(collector_class)
        if instance is None:
            raise RuntimeError(
                f"Internal error: Collector instance not found for classifier {classifier_class.__name__} in factory {type(self).__name__}",
            )

        instance.register(options)

        for dependency in instance.used_contexts():
            self.push_classifier(dependency, origin, list(stack))

    def pop_classifier(
        self,
        classifier: TRegisterEvaluationContext,
        origin: str,
        stack: list[type[CollectorClassifierBase[Any]]] | None = None,
    ) -> None:
        if stack is None:
            stack = []

        classifier_class: type[CollectorClassifierBase[Any]]
        options: dict[str, Any] | None = None

        if isinstance(classifier, dict):
            classifier_class = classifier["class"]
            options = classifier.get("options")
        else:
            classifier_class = classifier

        if classifier_class in stack:
            raise RuntimeError(
                f"Circular dependency detected in factory {type(self).__name__} "
                f"for classifier {classifier_class.__name__} (origin: {origin}, "
                f"stack: {[c.__name__ for c in stack]})",
            )

        stack.append(classifier_class)

        collector_class = self._collectors_per_classifier.get(classifier_class)
        if collector_class is None:
            raise RuntimeError(
                f'Factory "{type(self).__name__}" does not support classifier '
                f'"{classifier_class.__name__}" (origin: {origin}, '
                f"stack: {[c.__name__ for c in stack]})",
            )

        instance = self._collectors.get(collector_class)
        if instance is None:
            raise RuntimeError(
                f"Internal error: Collector instance not found for classifier {classifier_class.__name__} in factory {type(self).__name__}",
            )

        instance.unregister(options)

        for dependency in instance.used_contexts():
            self.pop_classifier(dependency, origin, list(stack))

    async def pre_notify(self, context: DataEvaluationContext, mediator: ObserverMediator) -> bool:
        from ..classifiers.dynamic_global_properties_classifier import DynamicGlobalPropertiesClassifier

        dgp = await context.get(DynamicGlobalPropertiesClassifier)
        head_block_number: int = dgp["head_block_number"]
        has_block_number_changed = self._current_block_number != head_block_number
        if has_block_number_changed:
            if self._current_block_number is not None:
                self._current_block_number += 1
            else:
                self._current_block_number = head_block_number
        return has_block_number_changed

    async def post_notify(self, context: DataEvaluationContext, mediator: ObserverMediator) -> None:
        pass

    def _rebuild_data_evaluation_context(self) -> DataEvaluationContext:
        context = DataEvaluationContext(self)
        for classifier, collector_class in self._collectors_per_classifier.items():
            instance = self._collectors.get(collector_class)
            if instance is None:
                raise RuntimeError(
                    f"Internal error: Collector instance not found for classifier {classifier.__name__} in factory {type(self).__name__}",
                )
            if not instance.has_registered:
                continue
            context.inject(classifier, instance)
        return context

    def collect(self) -> DataEvaluationContext:
        return self._rebuild_data_evaluation_context()
